### ISSUE TITLE
rule to enforce returning promise in async tests instead of done

### ISSUE THREAD

**Issue by zowers** [lo1tuma/eslint-plugin-mocha#62]
don't know if it is possible to enforce usage of promise returns, probably not.
at least it should be possible to check that `done` argument is not used and a `return something` is used instead

**Comment by lo1tuma** [comment]
Thanks for the issue.

What would be the intent of such a rule? Is it only about consistency or does it prevent some issues with the `done` callback`?

**Comment by zowers** [comment]
that would be used as a reminder of a better code made possible by the use of promises for all the project members.

most our code is promise-based except tests, which mostly `done` callbacks based, which happened due to lack of knowledge that mocha supports promises.

`done` callbacks are error-prone, it's easy to forget to catch errors and call done(err)

**Comment by lo1tuma** [comment]
Ok, sounds like it is a stylistic rule.I think we can do this by simply looking for the `done` argument and warn whenever it is specified. Checking if the `return` statement is present could be covered by #5.

**Comment by jfmengels** [comment]
I have been trying to write tests this way, and it works fine for normal cases, but I'm struggling to find a good way to find that a Promise should be rejected (with or without a check on the rejection reason). Any tips on that? I think tips for that should probably appear in the rule documentation.

**Comment by zowers** [comment]
promise should be rejected? what you mean?
the rule should disallow `done` callback in `it`/`beforeEach`/`afterEach`, and it doesn't have anything with promise state

**Comment by jfmengels** [comment]
I mean, how do you write a test where you expect a Promise to be rejected.

```js
it('should reject when no arguments are passed', function(done) {
  lib.foo()
    .then(function() {
      done(new Error('Expected error'));
    })
    .catch(function(error) {
      assert.ok(error);
      done();
    });
});
```

I have trouble writing tests like these withtout the `done` callback and without having the test pass when the promise is not rejected.

**Comment by lo1tuma** [comment]
```js
it('should reject when no arguments are passed', function() {
  return lib.foo()
    .then(function() {
      assert(false);
    })
    .catch(function(error) {
      assert.ok(error);
    });
});
```

I think that should work.

**Comment by zowers** [comment]
@jfmengels it's unrelated to the issue

you would use `chai-as-promised` (https://github.com/domenic/chai-as-promised)

**Comment by zowers** [comment]
or using `chai-as-promised`

```
it('should reject when no arguments are passed', function() {
  return expect( lib.foo() ).to.be.rejected;
});
```

**Comment by jfmengels** [comment]
@lo1tuma Unfortunately, your solution doesn't work, because the error thrown by the `assert.false()` statement would be catched by the following `catch`. You could then make new assertions in the catch to check what kind of error was thrown and make sure it was not caused by the `assert.false()` (assuming you know what kind of error could be thrown by the tested function).

I actually remembered a solution that I used, which uses the two arguments of `then`

```js
it('should reject when no arguments are passed', function() {
  return lib.foo()
    .then(function() {
      assert(false);
    }, function(error) {
      assert(error);
    });
});
```

This works. But I find using the two arguments instead of a `catch` a bit confusing and harder to read.

And yes, `chai-as-promised` works if you use `chai`, but I think this needs to be something that you can do without a special library.

**Comment by jfmengels** [comment]
Anyway, now that I have found a solution, I'm all for this rule, but the documentation should probably give tips on how to achieve this (normal case and my expected error case).

**Comment by lo1tuma** [comment]
This is already covered well enough by the existing rule set, so I do not think we should add a dedicated rule for it.

For projects that want promise-only async tests, this can already be expressed with:

```js
{
    rules: {
        "mocha/no-synchronous-tests": ["error", { allowed: ["promise"] }],
        "mocha/no-return-and-callback": "error"
    }
}
```

That gives the main behavior requested here without adding another overlapping rule.

Also, in modern TypeScript-heavy projects, this usually does not need Mocha-specific linting in the first place. If the test callback type is constrained correctly, using `done` vs returning a promise is often something the type system can help make explicit.

Closing this as not planned.
